### PR TITLE
Better semantic highlighting

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
image

image

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
It's finally ready,

**Comment by smoothdeveloper** [comment]
@vasily-kirichenko there is a bunch of commented code blocks, do you mind deleting those?

It seems that required signficant additions to the existing logic, thanks a lot for that work!

File: src/fsharp/vs/service.fs [hunk 1]
```diff
@@ -1418,25 +1432,47 @@ type TypeCheckInfo
...
+                Some (m, SemanticClassificationType.ComputationExpression)
+            | CNR(_, (Item.Value vref), _, _, _, _, m) when vref.IsMutable ->
+                Some (m, SemanticClassificationType.MutableVar)
+            | CNR(_, (Item.Value vref), _, _, _, _, m) when isFunction g vref.Type ->
+                if vref.IsPropertyGetterMethod || vref.IsPropertySetterMethod then
+                    Some (m, SemanticClassificationType.Property)
+                elif not (IsOperatorName vref.DisplayName) then
+                    Some (m, SemanticClassificationType.Function)
+                else None
+            //| CNR(_, (Item.Value vref), _, _, _, _, m) ->
```
Comment (KevinRansom) [inline src/fsharp/vs/service.fs]: Commented out ?

File: vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    [<Export(typeof<EditorFormatDefinition>)>]
+    [<ClassificationType(ClassificationTypeNames = FSharpClassificationTypes.Printf)>]
+    [<Name(FSharpClassificationTypes.Printf)>]
+    [<UserVisible(true)>]
+    [<Order(After = PredefinedClassificationTypeNames.String)>]
+    type internal FSharpPrintfTypeFormat() as self =
+        inherit ClassificationFormatDefinition()
+        
+        do self.DisplayName <- "F# Printf Format"
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs]: Localize ?

File: vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    [<Export(typeof<EditorFormatDefinition>)>]
+    [<ClassificationType(ClassificationTypeNames = FSharpClassificationTypes.MutableVar)>]
+    [<Name(FSharpClassificationTypes.MutableVar)>]
+    [<UserVisible(true)>]
+    [<Order(After = PredefinedClassificationTypeNames.Keyword)>]
+    type internal FSharpMutableVarTypeFormat() as self =
+        inherit ClassificationFormatDefinition()
+        
+        do self.DisplayName <- "F# Mutable Variables / Reference Cells"
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs]: Localize?

File: vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    [<Export(typeof<EditorFormatDefinition>)>]
+    [<ClassificationType(ClassificationTypeNames = FSharpClassificationTypes.Function)>]
+    [<Name(FSharpClassificationTypes.Function)>]
+    [<UserVisible(true)>]
+    [<Order(After = PredefinedClassificationTypeNames.Keyword)>]
+    type internal FSharpFunctionTypeFormat() as self =
+        inherit ClassificationFormatDefinition()
+        
+        do self.DisplayName <- "F# Functions / Methods"
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs]: Localize

File: vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    
+    [<Export(typeof<EditorFormatDefinition>)>]
+    [<ClassificationType(ClassificationTypeNames = FSharpClassificationTypes.Property)>]
+    [<Name(FSharpClassificationTypes.Property)>]
+    [<UserVisible(true)>]
+    [<Order(After = PredefinedClassificationTypeNames.Keyword)>]
+    type internal FSharpPropertyFormat() as self =
+        inherit ClassificationFormatDefinition()
+        
+        do self.DisplayName <- "F# Properties"
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs]: Localize

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 6]
```diff
@@ -618,4 +614,75 @@ module internal Extensions =
...
+                            args
+                                         
+                    // A simple overload resolution based on the name and number of parameters only
+                    let methods = 
+                        [ for m in methods do
+                            if m.Name = name && m.GetParameters().Length = args.Length then 
+                                yield m ]
+                                         
+                    match methods with
+                    | [] -> failwithf "No method '%s' with %d arguments found" name args.Length
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: Localize?

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 7]
```diff
@@ -618,4 +614,75 @@ module internal Extensions =
...
+                    | _ :: _ :: _ -> 
+                        failwithf "Multiple methods '%s' with %d arguments found" name args.Length
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: Localize?

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 8]
```diff
@@ -618,4 +614,75 @@ module internal Extensions =
...
+            |> unbox<'R>
+        else 
+            // When the 'o' object is 'System.Type', we access static properties
+            let typ, flags, instance = 
+                if (typeof<System.Type>).IsAssignableFrom(o.GetType()) then unbox o, staticFlags, null
+                else o.GetType(), instanceFlags, o
+            
+            // Find a property that we can call and get the value
+            let prop = typ.GetProperty(name, flags)
+            if prop = null then failwithf "Property '%s' not found in '%s' using flags '%A'." name typ.Name flags
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: Localize

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 9]
```diff
@@ -618,4 +614,75 @@ module internal Extensions =
...
+            let meth = prop.GetGetMethod(true)
+            if prop = null then failwithf "Property '%s' found, but doesn't have 'get' method." name
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: Localize

File: vsintegration/src/FSharp.LanguageService/Colorize.fs [hunk 10]
```diff
@@ -121,21 +121,21 @@ type internal FSharpScanner(makeLineTokenizer : string -> FSharpLineTokenizer) =
...
+            let color = colorInfo.ColorClass
                 // Upgrade identifiers to keywords based on extra info
-                match colorInfo.ColorClass with
-                | FSharpTokenColorKind.Identifier
-                | FSharpTokenColorKind.UpperIdentifier ->
-                    match tryFindExtraInfo (line, colorInfo.LeftColumn, colorInfo.RightColumn) with
-                    | None -> FSharpTokenColorKind.Identifier
-                    | Some info -> info // extra info found
-                | c -> c
+                //match colorInfo.ColorClass with
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.LanguageService/Colorize.fs]: Commented out text?

**Comment by KevinRansom** [comment]
@vasily-kirichenko sounds good.

**Comment by KevinRansom** [comment]
Thank you for this.

**Comment by cartermp** [comment]
🎉

**Comment by KevinRansom** [comment]
Thank you for this.

**Comment by dsyme** [comment]
@vasily-kirichenko 

> About the commented code in service.fs, it's an attempt to determine if vref is of type Ref'1. I asked @dsyme for help, but he's not answered.

Please ping me again (feel free to drop me a DM on twitter if I miss a message)

**Comment by vasily-kirichenko** [comment]
@dsyme https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/vs/service.fs#L1445-L1454
I've added supporting stuff into `TcGlobals.fs`, see https://github.com/Microsoft/visualfsharp/pull/2348/files#diff-07058ef42dbcab9137d8a2f987f63416R606
Maybe I'm doing it all wrong.

File: vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    [<Export(typeof<EditorFormatDefinition>)>]
+    [<ClassificationType(ClassificationTypeNames = FSharpClassificationTypes.Function)>]
+    [<Name(FSharpClassificationTypes.Function)>]
+    [<UserVisible(true)>]
+    [<Order(After = PredefinedClassificationTypeNames.Keyword)>]
+    type internal FSharpFunctionTypeFormat() as self =
+        inherit ClassificationFormatDefinition()
+        
+        do self.DisplayName <- SR.FSharpFunctionsOrMethodsClassificationType.Value
+           self.ForegroundColor <- Nullable Colors.Black
```
Comment (majocha) [inline vsintegration/src/FSharp.Editor/Classification/ClassificationDefinitions.fs]: Looks like this causes text to be black in dark theme.
It seems that leaving this unassigned results in automatic color (black for light themes, white for dark) that still can be customized.

**Comment by vasily-kirichenko** [comment]
@majocha themes tracking exists in VFPT (it has small bugs though and reflection is used and...) and was in initial PR here, but tests failed with type load exception. I'd spent several days trying to fix it, but failed and removed the themes support. If anybody who uses the dark theme wants to finish it, he is welcome.

**Comment by majocha** [comment]
Thanks @vasily-kirichenko!  Sorry I didn't read the whole PR before posting, I was just excited that I can build from master again and this was something I noticed first. I'll try dig into it.

**Pull request merged**